File: windows/GameBox.Windows.Core/CompanionProtocol.cs
using System.Security.Cryptography;
using System.Text;

namespace GameBox.Windows.Core;

/// <summary>
/// Versioned, authenticated request primitives shared by the Windows companion's
/// future device client. Pairing secrets are never serialized with the device profile.
/// </summary>
public static class CompanionProtocol
{
    public const int Version = 1;
    public const string AuthorizationHeader = "X-GameBox-Authorization";

    public static string CreatePairingSecret()
        => Convert.ToHexString(RandomNumberGenerator.GetBytes(32)).ToLowerInvariant();

    public static string CreateAuthorization(string pairingSecret, string method, string requestPath, long unixTimeSeconds)
    {
        if (!IsValidSecret(pairingSecret)) throw new ArgumentException("Pairing secret is invalid.", nameof(pairingSecret));
        if (string.IsNullOrWhiteSpace(method)) throw new ArgumentException("Method is required.", nameof(method));
        if (!requestPath.StartsWith('/', StringComparison.Ordinal) || requestPath.Contains("..", StringComparison.Ordinal))
            throw new ArgumentException("Request path must be absolute and traversal-free.", nameof(requestPath));
        if (unixTimeSeconds <= 0) throw new ArgumentOutOfRangeException(nameof(unixTimeSeconds));

        var payload = $"v{Version}\n{method.Trim().ToUpperInvariant()}\n{requestPath}\n{unixTimeSeconds}";
        var key = Convert.FromHexString(pairingSecret);
        var hash = HMACSHA256.HashData(key, Encoding.UTF8.GetBytes(payload));
        return $"v{Version}:{unixTimeSeconds}:{Convert.ToHexString(hash).ToLowerInvariant()}";
    }

    public static bool VerifyAuthorization(
        string pairingSecret,
        string method,
        string requestPath,
        string? authorization,
        long nowUnixTimeSeconds,
        long allowedSkewSeconds = 120)
    {
        if (!IsValidSecret(pairingSecret) || string.IsNullOrWhiteSpace(authorization) || allowedSkewSeconds < 0) return false;
        var parts = authorization.Split(':');
        if (parts.Length != 3 || parts[0] != $"v{Version}" || !long.TryParse(parts[1], out var timestamp)) return false;
        if (Math.Abs(nowUnixTimeSeconds - timestamp) > allowedSkewSeconds) return false;
        try
        {
            var expected = CreateAuthorization(pairingSecret, method, requestPath, timestamp);
            var expectedBytes = Encoding.UTF8.GetBytes(expected);
            var actualBytes = Encoding.UTF8.GetBytes(authorization);
            return expectedBytes.Length == actualBytes.Length && CryptographicOperations.FixedTimeEquals(expectedBytes, actualBytes);
        }
        catch (ArgumentException) { return false; }
    }

    private static bool IsValidSecret(string value) =>
        value.Length == 64 && value.All(static c => char.IsAsciiHexDigit(c));
}

File: app/src/main/java/com/gamebox/os/companion/CompanionProtocol.kt
package com.gamebox.os.companion

import java.nio.charset.StandardCharsets
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Shared protocol-v1 verification for a future paired Windows companion endpoint. */
object CompanionProtocol {
    const val VERSION = 1
    const val AUTHORIZATION_HEADER = "X-GameBox-Authorization"

    fun createAuthorization(secret: String, method: String, requestPath: String, unixTimeSeconds: Long): String {
        require(validSecret(secret)) { "Pairing secret is invalid" }
        require(method.isNotBlank()) { "Method is required" }
        require(requestPath.startsWith('/') && !requestPath.contains("..")) {
            "Request path must be absolute and traversal-free"
        }
        require(unixTimeSeconds > 0) { "Timestamp is required" }
        val payload = "v$VERSION\n${method.trim().uppercase()}\n$requestPath\n$unixTimeSeconds"
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(hexToBytes(secret), "HmacSHA256"))
        return "v$VERSION:$unixTimeSeconds:${mac.doFinal(payload.toByteArray(StandardCharsets.UTF_8)).toHex()}"
    }

    fun verifyAuthorization(
        secret: String,
        method: String,
        requestPath: String,
        authorization: String?,
        nowUnixTimeSeconds: Long,
        allowedSkewSeconds: Long = 120,
    ): Boolean {
        if (!validSecret(secret) || authorization == null || allowedSkewSeconds < 0) return false
        val parts = authorization.split(':')
        if (parts.size != 3 || parts[0] != "v$VERSION") return false
        val timestamp = parts[1].toLongOrNull() ?: return false
        if (kotlin.math.abs(nowUnixTimeSeconds - timestamp) > allowedSkewSeconds) return false
        return runCatching {
            val expected = createAuthorization(secret, method, requestPath, timestamp)
            constantTimeEquals(expected, authorization)
        }.getOrDefault(false)
    }

    private fun validSecret(value: String): Boolean = value.length == 64 && value.all { it.isDigit() || it.lowercaseChar() in 'a'..'f' }

    private fun hexToBytes(value: String): ByteArray = ByteArray(value.length / 2) { index ->
        value.substring(index * 2, index * 2 + 2).toInt(16).toByte()
    }

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

    private fun constantTimeEquals(left: String, right: String): Boolean {
        val leftBytes = left.toByteArray(StandardCharsets.UTF_8)
        val rightBytes = right.toByteArray(StandardCharsets.UTF_8)
        if (leftBytes.size != rightBytes.size) return false
        var difference = 0
        leftBytes.indices.forEach { index -> difference = difference or (leftBytes[index].toInt() xor rightBytes[index].toInt()) }
        return difference == 0
    }
}

